File: src/algorithms/trees/advanced/n-ary-tree-traversal/sources/n-ary-tree-traversal.py
# N-ary Tree Traversal — preorder visit using children list

class NAryNode:
    def __init__(self, value: int, children=None):
        self.value = value
        self.children = children or []

def n_ary_tree_traversal(root) -> list:
    result = []  # @step:initialize

    def preorder(node):
        if not node:
            return  # @step:initialize

        result.append(node.value)  # @step:visit

        for child in node.children:
            preorder(child)  # @step:traverse-next

    preorder(root)  # @step:initialize
    return result  # @step:complete

File: src/algorithms/trees/advanced/n-ary-tree-traversal/sources/NAryTreeTraversal.cpp
// N-ary Tree Traversal — preorder visit using children vector
#include <vector>
using namespace std;

struct NAryNode {
    int value;
    vector<NAryNode*> children;
    NAryNode(int v) : value(v) {}
};

void preorder(NAryNode* node, vector<int>& result) {
    if (!node) return; // @step:initialize

    result.push_back(node->value); // @step:visit

    for (NAryNode* child : node->children) {
        preorder(child, result); // @step:traverse-next
    }
}

vector<int> nAryTreeTraversal(NAryNode* root) {
    vector<int> result; // @step:initialize

    preorder(root, result); // @step:initialize
    return result; // @step:complete
}
